cpp/include/cudf/detail/copy_if_else.cuh

```diff
@@ -0,0 +1,225 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cudf/cudf.h>
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_view.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/utilities/type_dispatcher.hpp>
+#include <cudf/detail/utilities/cuda.cuh>
+#include <utilities/bit_util.cuh>
+#include <utilities/cuda_utils.hpp>
+
+#include <cub/cub.cuh>
+
+
+namespace cudf {
+namespace experimental {
+namespace detail {
+
+namespace {    // anonymous
+
+template <typename T, typename Filter, bool has_validity>
+__global__
+void copy_if_else_kernel(  column_device_view const lhs,
+                           column_device_view const rhs,
+                           Filter filter,
+                           mutable_column_device_view out,
+                           cudf::size_type * __restrict__ const null_count)
+{   
+   const cudf::size_type tid = threadIdx.x + blockIdx.x * blockDim.x;   
+   const int w_id = tid / warp_size;
+   // begin/end indices for the column data
+   cudf::size_type begin = 0;
+   cudf::size_type end = lhs.size();   
+   // warp indices.  since 1 warp == 32 threads == sizeof(bit_mask_t) * 8,
+   // each warp will process one (32 bit) of the validity mask via 
+   // __ballot_sync()
+   cudf::size_type w_begin = cudf::util::detail::bit_container_index<bit_mask::bit_mask_t>(begin);
+   cudf::size_type w_end = cudf::util::detail::bit_container_index<bit_mask::bit_mask_t>(end);
+
+   // lane id within the current warp
+   const int w_lane_id = threadIdx.x % warp_size;
+
+   // store a null count for each warp in the block
+   constexpr cudf::size_type b_max_warps = 32;
+   __shared__ uint32_t b_warp_null_count[b_max_warps];
+   // initialize count to 0. we have to do this because the WarpReduce
+   // at the end will end up summing all values, even ones which we never 
+   // visit.
+   if(has_validity){   
+      if(threadIdx.x < b_max_warps){
+         b_warp_null_count[threadIdx.x] = 0;
+      }   
+      __syncthreads();   
+   }   
+
+   // current warp.
+   cudf::size_type w_cur = w_begin + w_id;         
+   // process each grid
+   while(w_cur <= w_end){
+      // absolute element index
+      cudf::size_type index = (w_cur * warp_size) + w_lane_id;
+      bool in_range = (index >= begin && index < end);
+
+      bool valid = true;
+      if(has_validity){
+         valid = in_range && filter(index) ? lhs.is_valid(index) : rhs.is_valid(index);
+      }
+
+      // do the copy if-else, but only if this element is valid in the column to be copied 
+      if(in_range && valid){ 
+         out.element<T>(index) = filter(index) ? lhs.element<T>(index) : rhs.element<T>(index);
+      }
+      
+      // update validity
+      if(has_validity){
+         // get mask indicating which threads in the warp are actually in range
+         int w_active_mask = __ballot_sync(0xFFFFFFFF, in_range);      
```

Comment:
I think you need to compute and update the active threads wrt the while loop. See other comments for additional changes.
```suggestion
         int w_active_mask = __ballot_sync(active_threads, in_range);      
```

Reply:
Is this to keep the logic of the null counting simpler, or is there an actual bug here?  This loop is modelled pretty much straight off of copy_range.

Reply:
See above comment. The way you've structured the loop this isn't necessary (but it is necessary in more traditional grid stride loops).

Reply:
Mark - final verdict? Leave as is or add the tweak?

Reply:
Interestingly, as fate would have it I was looking into https://github.com/rapidsai/cudf/pull/2842 and I'm 99% sure the solution for that is the exact same logic as this (ie, the valid_if way of doing it).  So, performance issues aside (the valid_if way causes __ballot_sync calls even when there's no validity data) I guess it makes sense to use the same pattern for all 3 of these (copy_if_else, valid_if and the rolling_window bug) for the sake of consistency.

Reply:
It's fine as-is as Jake points out.